server/src/lsp-server.ts

```diff
@@ -75,8 +75,11 @@ export class LspServer {
         if (this.options.tsserverPath) {
             return this.options.tsserverPath;
         }
-        // 1) look into node_modules of workspace root
-        const executable = findPathToModule(this.rootPath(), `.bin/${getTsserverExecutable()}`);
+        // 1) look into node_modules or yarn PnP SDKs of workspace root
+        const executable = findPathToModule(this.rootPath(), [
+            path.join("node_modules", ".bin", getTsserverExecutable()),
+            path.join(".yarn", "sdks", "typescript", "lib", "tsserver.js"),
+        ])
```

Comment:
We're likely gonna get https://github.com/theia-ide/typescript-language-server/pull/195 in since that one is clean and a proper way to do it. That might force you to rethink this part but I think it should be anyway because this is just too hacky. We really should be relying on the native resolver first and then add specific behavior for yarn sdk if really necessary.

Reply:
Ok, happy to look at that. I guess we could add a new step 1 for resolving the tsserver path that checks for the yarn sdk first. I think it will need to be that way round as `require.resolve` will work for PnP but it will find the wrong file. There could be a slight performance penalty with that approach for non-yarn PnP project as they'll check for `.yarn/sdks` all the way up to `/` before using `require.resolve`.

Reply:
We don't really need to exist `findPathToModule`. This logic can be separate. This only needs to be done on initialize, from one place, anyway.

That said, `findPathToModule` is only used from two places during initialize anyway so either way is fine now right now at least.

Reply:
Can't it use something like the `tsserver.tsdk` config that is generated for `coc-vim` and `vscode`?
https://github.com/yarnpkg/berry/blob/a6312e23c2cff2ee8bc338a9eaff486aa5cb97bc/.vim/coc-settings.json#L5

Reply:
I guess we could go searching through these paths for each editor in the language server (`.vim/coc-settings.json` and `.vscode/settings.json`) or editor could do it themselves and pass through the path but then each editor / user of the language server would have to solve yarn PnP

Reply:
Reading editor-specific configuration files is very much against the idea of LSP. I've very much opposed to that.

Also, nobody uses this server with vscode since VSCode has built-in server so I'm not sure it would be even relevant to use the option defined there.

Reply:
This part was changed so resolving this issue.

Reply:
That said, I agree that ideally the editor would do that part. Currently it's possible to pass the tsserver path through command line option (https://github.com/theia-ide/typescript-language-server/blob/49520504d5dc2b808c8ffce9457bbfa1129315b1/server/src/cli.ts#L24-L24)